docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+
+## Possible solution
+
+The main idea is to give the current connection holder (as indicated from the Function), the edge to
+continue processing messages for this device. The performance of that gateway will not be impacted.
+
+A local (per LNS) in-memory dictionary will be used for that purpose to map DevEuis to a flag
+indicating whether the LNS is the preferred gateway or not.
+
+### Handling of cache refresh
+
+When we create the (singleton) instance of the NetworkServer.LoRaDeviceCache, we start a background
+periodical task to ensure the device twins for all the devices that connected to that LNS are kept
+fresh. In the case where we need to get a device twin, this could trigger a connection ping-pong.
+
+#### Decision
+
+For now we don't make changes here and accept there will be a potential connection switching
+periodically (currently every 2 days if the device was not refreshed in the meantime). This will be
+revisited with either [this issue about the frame counter
+  drift](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1499)
+ (one of the solutions there is to fetch twin and close connection immediately) or [this issue
+ specific to cache refresh](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1505)
+
+#### Alternatives
+
+- Contact the Function to see whether we are the owning gateway. If we are, refresh the twin (and
+  keep the connection open). If we are not:
+  - We adjust the LastSeen property but not actually refresh the entry in the cache.
+    When the next data message for this devices comes in and an entry is in cache for such device, in the
+    event that the Azure Function is marking our LNS as the new "winning" gateway, we have a stale
+    twin in the cache that needs to be updated before validating the request (see [main data flow](#main-data-message-flow)).
+  - Alternatively, we remove device entry from the cache. This idea is discarded because if we were
+    to do this, a "get twin" operation would be triggered as soon as the next data message is coming
+    in from the device in question (for the resolution of devAddr).
+  - We still refresh the twin and close the connection immediately.
+- Remove the background refresh feature all-together: discarded because solutions are possible and
+  not overly complex.
+
+### Handling of Join requests
+
+Join requests in isolation currently do not have the connection stealing issue, as they already rely on the
+Function for the DevNonce check. If the current LNS is not the preferred gateway, it drops the
+message immediately.
+
+Problem
+
+- Join received from LNS1 and LNS2: LNS1 wins
+- First data message received only on LNS2, Function should inform LNS1 that it's the losing one
+  (information about the owning LNS needs to be "shared" between joins and data messages).
+
+#### Decision
+
+We decide to go with the simpler solution of closing the connection immediately after a Join and let
```

Comment:
The decision to close the connection is not only for the join request, but for any initialization operation. When we load the device into the cache for the first time, we need to load the twins. No matter if it is a join or not, we need to close the connection, because there is no "leader election pattern" on the load operations. We do not know if we own the connection after that operation, potentially leaving the connection open. Only data message flow, does determine the ownership of the connection and maintains it.

Reply:
Agree. Not sure if you request a change here?

If you would like the wording to be more generic: this section is specifically for the joins but I made sure to clarify that we close the connection also on the background refresh and cache initialization cases.

Reply:
All good - was referring to the close when initializing for a data message, which is covered later.